### ISSUE TITLE
[HOLD for payment 2024-01-31] [HOLD for payment 2024-01-24] [$500] Feature Request: Three-finger swipe to go back and forward doesn't work on desktop

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#33435]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:**
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @rafecolton 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1703008037476039

## Action Performed:

1. Swipe on the trackpad with three fingers (or whatever you have configured in your system settings - could be two).

## Expected Result:
 Navigates back and forward like in the browser or when using cmd + [ or cmd + ]

## Actual Result:
Does nothing

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01fce9f35e954f949c</li>
        <li>Upwork Job ID: 1742359014682963968</li>
        <li>Last Price Increase: 2024-01-09</li>
<li>Automatic offers: </li>
<ul>
<li>shubham1206agra | Reviewer | 28092901</li>
<li>gijoe0295 | Contributor | 28092902</li>
</ul></li>
    </ul>
</details>

**Comment by mallenexpensify** [comment]
A lil unsure what to do here.  Our concept of navigation is _weird_.  We don't have back/forward buttons on Desktop and `command+[` and `]` work for shortcut keys but we don't currently support other keyboard shortcuts.  

@shubham1206agra , what do you think we should do?

**Comment by shubham1206agra** [comment]
I think we should not do anything for swipe gesture as this is browser specific feature (only safari supports this).
And for navigation part, at best, we should show forward / back button for navigation (not sure), or add the navigation shortcut in keyboard shortcut screen so the user can be informed.

**Comment by shubham1206agra** [comment]
> Most macOS trackpads are not configured to allow this kind of swiping anymore

**Comment by shubham1206agra** [comment]
Tbh I don't think anybody will use swiping (that too 3 finger).

**Comment by rafecolton** [comment]
I use it every day. It is supported by default in every browser I've used and on other desktop apps like Slack.

**Comment by mallenexpensify** [comment]
Thanks all for the feedback! For Desktop, it the fix as simple as
Mac - For forward/back navigation, default to MacOS settings

and, for PC, is there anything we want to do now, so that if/when we build a PC Desktop app, we also default to any forward/back navigation in settings there?

**Comment by mallenexpensify** [comment]
Was just testing swiping on new.expensify.com in Chrome and wasn't able to use my default 'swipe two fingers' to go forward/back, so the issue might not only be related to Desktop

**Comment by shubham1206agra** [comment]
@mallenexpensify If you feel we need to implement swipe gesture on Desktop, please make the issue external.

**Comment by shubham1206agra** [comment]
@mallenexpensify Bump on above ^

**Comment by mallenexpensify** [comment]
I'm unsure of the technical details for how this should be best accomplished.  From a non-engineer point of view, whatever is selected in Trackpad settings on mac for "Swipe between pages" below should navigate Desktop back/forward the same way that's currently possible via `cmd+[` and `]` 

<img width="520" alt="image" src="https://github.com/Expensify/App/assets/22508304/d8aaedd8-2fe8-4905-8b3b-f3a17521d1f5">

cc @rafecolton to make sure this _correctly_ accomplishes what you're looking for (and in a way that not unaligned with other aspects/part of the app/code)

**Comment by mallenexpensify** [comment]
@shubham1206agra can you review the existing proposals and comments above and provide an update?  I like the term "support native navigation gestures", thanks @rafecolton

**Comment by shubham1206agra** [comment]
@gijoe0295's [proposal](https://github.com/Expensify/App/issues/33435#issuecomment-1866748574) works fine for me. Lets go with this.

🎀 👀 🎀 C+ Reviewed

**Comment by mallenexpensify** [comment]
@rafecolton, can you test this and let me know if it's working for you? It doesn't appear to be for me (but I could be holding it wrong).   PR hit production two days ago.

**Comment by rafecolton** [comment]
It seems to be... backwards? Here's what I did:

1. Navigate to Chronos
2. Quit and reopen NewDot
3. Navigate to #social => DM => #social => thread in #social

Three fingers to the left (back) does nothing. However, three fingers to the right (forward) goes back through the list! Once I'm back at Chronos, swiping left (back) takes me back through.

So the gestures are supported-ish, but there is a bug either in how we order the nav history or maybe the swipe gestures are swapped (back <=> forward).

**Comment by rafecolton** [comment]
Hm that's really interesting. It could be a locale thing? Like I don't see any way to invert them in my system settings, I only see this:

![Screenshot 2024-01-19 at 12 06 08 PM](https://github.com/Expensify/App/assets/1058475/26f957b7-76d3-47ea-90e4-2d5bdc68fb21)

(I have it set to three-fingers only because two fingers does horizontal scrolling.) I'm not sure if it's possible, but ideally it should not be hardcoded either way and should instead map to the system gestures. So whichever direction is back on other apps on your system (e.g. browsers), it should be the same on NewDot.

**Comment by mallenexpensify** [comment]
Desktop, Version 1.4.28-0 , isn't navigating all all, back or forward, when i use two fingers to swipe left or right
<img width="491" alt="image" src="https://github.com/Expensify/App/assets/22508304/f6475924-cb83-428f-8004-d2df742696cf">

**Comment by shubham1206agra** [comment]
@rafecolton Slack doesn't allow the same either.

**Comment by rafecolton** [comment]
Hm, interesting. I tried changing the settings to two or three fingers, and three fingers still works in slack, but two fingers does not. _However_, after changing it, two finger swipe _does_ work in browsers.

**Comment by mallenexpensify** [comment]
`TIL` that 2 fingers doesn't work (ever if that's your mac setting) and 3 fingers does, both for Slack and now NewDot.  
@rafecolton can you confirm that 3 fingers is working for you for NewDot?  Thx

**Comment by rafecolton** [comment]
Yep, seems to be working as expected now

**Comment by mallenexpensify** [comment]
@shubham1206agra can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~01fce9f35e954f949c

@gijoe0295 , please confirm your name in Upwork so I can ensure I'm paying the right person?
Thx

**Comment by shubham1206agra** [comment]
**BugZero Checklist:** The PR adding this new feature has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

* [x]  [@shubham1206agra] Please propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure  the new feature will work correctly on production in further releases.

Steps on Desktop only
1. Swipe on the trackpad with three fingers to navigate back and forward like in the browser or when using cmd + [ or cmd + ]
* [ ]  [@mallenexpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon.

**Comment by shubham1206agra** [comment]
Offer accepted.

**Comment by shubham1206agra** [comment]
Not sure how this is related to our change. But will look into this.
@rafecolton Can you just in case list all ways to navigate in desktop?

**Comment by rafecolton** [comment]
> Not sure how this is related to our change. But will look into this.

Not sure if it's related mechanically, only flagged it because it's related conceptually. So appreciate you checking it out.

> Can you just in case list all ways to navigate in desktop?

I cannot, I don't know what they all are 😄 Off the top of my head, there should be three-finger swipe, cmd + left/right arrow, and cmd + square bracket.

**Comment by shubham1206agra** [comment]
I confirmed this. And this is not related to desktop only. This is also a problem with web too. And github and slack does the navigation on cmd + left/right arrow on web too.

So this is not a regression from this issue.

cc @rafecolton

**Comment by rafecolton** [comment]
@gijoe0295 not sure what you are referring to, there is no channel in question here. Are you sure this is the issue you meant to comment on?

@shubham1206agra thanks for checking. I don't know enough to say if it's a regression from this issue. Did you identify where the issue occurred?

You are not _quite_ correct about this not working on web. It works IF you are not focused on an input field. If you're in an input field, it goes to the beginning and end of the line. But if you click out of the input field, it still works in the browser because the browser supports back and forward using cmd + arrows. On desktop, it does not work at all.

**Comment by rafecolton** [comment]
Ah, that's just where I posted the concern in our bug-tracking channel, there's no new information there that does not appear in this GH

**Comment by rafecolton** [comment]
Thanks @amyevans, I updated the Slack thread to let everybody know it's a separate issue

**Comment by mallenexpensify** [comment]
Contributor: @gijoe0295 paid $500 via Upwork
Contributor+: @shubham1206agra paid $500 via Upwork.  

@gijoe0295 if you'd like to join the Slack channel, details are here
https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#slack-channels

I created the TestRail update steps
https://github.com/Expensify/Expensify/issues/366364
